File: solutions/go/queen-attack/3/queen_attack.go
package queenattack

import "errors"

var (
    ErrInvalidFormat = errors.New("Input format is invalid")
)

func CanQueenAttack(whitePosition, blackPosition string) (bool, error) {
    if whitePosition == blackPosition {
        return false, ErrInvalidFormat
    }
	err := errors.Join(validatePosition(whitePosition), validatePosition(blackPosition))
    if err != nil {
        return false, err
    }
    rowDiff, colDiff := whitePosition[0] - blackPosition[0], whitePosition[1] - blackPosition[1]
    return rowDiff == 0 || colDiff == 0 || rowDiff + colDiff == 0 || rowDiff - colDiff == 0, nil
}

func validatePosition(position string) error {
    if len(position) != 2 || !('a' <= position[0] && position[0] <= 'h') || !('1' <= position[1] && position[1] <= '8') {
        return ErrInvalidFormat
    }
    return nil
}

File: solutions/typescript/queen-attack/1/queen-attack.ts
type Position = readonly [number, number]

type Positions = {
  white: Position
  black: Position
}

export class QueenAttack {
  public readonly black: Position
  public readonly white: Position

  constructor({ white = [7, 3], black = [0, 3] }: Partial<Positions> = {}) {
    QueenAttack.validatePosition(white)
    QueenAttack.validatePosition(black)
    QueenAttack.validateDuplicate(white, black)
    this.white = white
    this.black = black
  }

  toString(): string {
    const board = Array.from({length: 8}, () => Array.from({length: 8}).fill('_'))
    board[this.white[0]][this.white[1]] = 'W'
    board[this.black[0]][this.black[1]] = 'B'
    return board.map(row => row.join(' ')).join('\n')
  }

  get canAttack(): boolean {
    const rowDiff = this.white[0] - this.black[0]
    const colDiff = this.white[1] - this.black[1]
    return rowDiff * colDiff * (rowDiff + colDiff) * (rowDiff - colDiff) === 0
  }

  static validatePosition(position: Position): void {
    for (const coord of position) {
      if (coord < 0 || coord >= 8) {
        throw new Error('Queen must be placed on the board')
      }
    }
  }

  static validateDuplicate(position1: Position, position2: Position): void {
    if (position1[0] === position2[0] && position1[1] === position2[1]) {
      throw new Error('Queens cannot share the same space')
    }
  }
}